### PR TITLE
Multiple files

### PR DESCRIPTION
**Pull request by rebinskiy**
Working version of a multiple files.

Missing: 
- Single client queue.
- Timeout.
- Junit.

### CODE REVIEW

**Comment by VincenzoFerme** [comment]
I do not see the following feedback implemented:

- https://github.com/ASQ-USI/asq-java-q-backend/pull/3#discussion_r91838421
- https://github.com/ASQ-USI/asq-java-q-backend/pull/3#discussion_r91838458

**Comment by rebinskiy** [comment]
@VincenzoFerme Ok, sorry forgot about it, now should be implemented.

**Comment by VincenzoFerme** [comment]
Where is this defined in the `package.json`?

**Comment by rebinskiy** [comment]
Now it is (push lag)

**Review by VincenzoFerme — APPROVED** [review]

File: app.js [hunk 1]
```diff
@@ -1,36 +1,109 @@
...
-    fs.createWriteStream(filePath).write(data);
+    const makeTarAndRun = () => {
 
-    tar.pack(dirPath).pipe(fs.createWriteStream(tarPath));
+        tar.pack(dirPath).pipe(fs.createWriteStream(tarPath));
+        javaBox.emit('runJava', clientId, main, tarPath);
+    };
 
-    javaBox.emit('runJava', socket, tarPath);
+    const writeFiles = () => {
```
Comment (triglian) [inline app.js:26]: How do you know when the file operations that `writeFiles()` performs have ended executing? You have no callbacks  (and any error handling in them)

File: javaBox.js [hunk 2]
```diff
@@ -5,80 +5,43 @@ const Docker = require('dockerode');
...
+    const souceLocation = tarPath;
+    const execution = dockerCommand(javacCmd, dockerCommand(javaCmd));
 
-    createJContainer(souceLocation, execution);
-}
+    createJContainer(clientId, souceLocation, execution);
+};
 
 
 // Creates and start a container with bash, JDK SE and more
```
Comment (triglian) [inline javaBox.js]: Creates and starts

File: javaBox.js [hunk 3]
```diff
@@ -5,80 +5,43 @@ const Docker = require('dockerode');
...
-        container.start(opts, (err, data) => {
+        const startOpts = {};
+        container.start(startOpts, (err, data) => {
 
-            let opts = {path: 'home'};
-            container.putArchive(javaSourceTar, opts, (err, data) => {
+            const tarOpts = {path: 'home'};
+            container.putArchive(javaSourceTar, tarOpts, (err, data) => {
 
                 callback(container);
```
Comment (triglian) [inline javaBox.js]: The first argument in callbacks in node.js should __always__ be the error. So this means that you should change this to `callback(null, container)`

File: javaBox.js [hunk 4]
```diff
@@ -91,40 +54,77 @@ function createJContainer(javaSourceTar, callback) {
...
 };
 
+function sendResponse() {
+
+    const feedback = {
+        clientId: container.clientId,
+        passed: false,
+        output: '',
+        errorMessage: stream.read().toString(),
+        //errorMessage: stream.read().toString().replace(/\u0000|\u0001/g, '').trim(),
```
Comment (triglian) [inline javaBox.js]: delete if not needed

File: app.js (hunk 1 above)
Comment (rebinskiy) [reply app.js:26]: fs.createWriteStream(filePath).write(file.data, fileAdded); at line 59 and fs.unlink(`${dirPath}/${oldFile}`, fileDeleted); at line 68 have fileAdded and fileDeleted as callbacks, these two decrease filesToDelete or filesToAdd int variables respectively and both call tryTarAndRun, that checks whether filesToDelete and filesToAdd are both 0 and, if they are, call makeTarAndRun

**Pull request merged**